### PR TITLE
BIA 565 - combine the studentAssessment and StudentAssessmentObjectiveFact

### PR DESCRIPTION
**Pull request by NSD-Ruben**
combine the studentAssessment and StudentAssessmentObjectiveFact

### CODE REVIEW

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        CASE WHEN SASOA.AssessmentIdentifier IS NULL THEN NULL ELSE CONCAT(SASOA.AssessmentIdentifier, '-', SASOA.IdentificationCode, '-', SASOA.Namespace) END AS ObjectiveAssessmentKey,
+		CONCAT(Assessment.AssessmentIdentifier, '-', StudentAssessment.Namespace) AS AssessmentKey,
+        Assessment.AssessmentIdentifier,
+        StudentAssessment.Namespace,
+        StudentAssessment.StudentAssessmentIdentifier,
+        StudentAssessment.StudentUSI,
+        CONCAT(Student.StudentUniqueId, '-', StudentSchoolAssociation.SchoolId) AS StudentSchoolKey,
+        School.SchoolId AS SchoolKey,
+        CONVERT(varchar, StudentAssessment.AdministrationDate, 112) as AdministrationDate,
+        COALESCE(ISNULL(StudentAssessmentScoreResult.Result,SASOASR.Result),'') AS StudentScore,
```
Comment (DavidJGapCR) [inline src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:26]: Why do we have a ISNULL inside a Coalesce? Maybe we can remove the ISNULL statement. Unless I am not seeing something.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+BEGIN
+	DROP VIEW analytics.asmt_StudentAssessmentFact
+END
+GO
+
+CREATE VIEW analytics.asmt_StudentAssessmentFact AS
+
+	SELECT
+        CONCAT(StudentAssessment.AssessmentIdentifier, '-', StudentAssessment.Namespace, '-', StudentAssessment.StudentAssessmentIdentifier, '-', StudentAssessment.StudentUSI) AS StudentAssessmentKey,
+		CASE WHEN SASOA.StudentUSI IS NULL THEN NULL ELSE CONCAT(SASOA.StudentUSI, '-', SASOA.IdentificationCode, '-', SASOA.AssessmentIdentifier, '-', SASOA.StudentAssessmentIdentifier, '-', SASOA.Namespace) END AS StudentObjectiveAssessmentKey,
```
Comment (DavidJGapCR) [inline src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:16]: Can you format the case statement a little bit better to make it more readable? There are other views where we have case statements as well, maybe you can find them and use them as a reference.

File: src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		CASE WHEN SASOA.StudentUSI IS NULL THEN NULL ELSE CONCAT(SASOA.StudentUSI, '-', SASOA.IdentificationCode, '-', SASOA.AssessmentIdentifier, '-', SASOA.StudentAssessmentIdentifier, '-', SASOA.Namespace) END AS StudentObjectiveAssessmentKey,
+        CASE WHEN SASOA.AssessmentIdentifier IS NULL THEN NULL ELSE CONCAT(SASOA.AssessmentIdentifier, '-', SASOA.IdentificationCode, '-', SASOA.Namespace) END AS ObjectiveAssessmentKey,
+		CONCAT(Assessment.AssessmentIdentifier, '-', StudentAssessment.Namespace) AS AssessmentKey,
+        Assessment.AssessmentIdentifier,
+        StudentAssessment.Namespace,
+        StudentAssessment.StudentAssessmentIdentifier,
+        StudentAssessment.StudentUSI,
+        CONCAT(Student.StudentUniqueId, '-', StudentSchoolAssociation.SchoolId) AS StudentSchoolKey,
+        School.SchoolId AS SchoolKey,
+        CONVERT(varchar, StudentAssessment.AdministrationDate, 112) as AdministrationDate,
```
Comment (DavidJGapCR) [inline src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql:20]: This line changes a lot compared to the SQL Server version. Take a look at the previous version.

File: src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        COALESCE(ISNULL(StudentAssessmentScoreResult.Result,SASOASR.Result),'') AS StudentScore,
```
Comment (DavidJGapCR) [inline src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql:21]: I don't think ISNULL exists in Postgres.

File: src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+            AND
+            SASOASR.Namespace = SASOAPL.Namespace
+            AND
+            SASOASR.AssessmentReportingMethodDescriptorId = SASOAPL.AssessmentReportingMethodDescriptorId
+	LEFT JOIN
+        edfi.Descriptor AS PerformanceLevelDescriptorObj ON
+            PerformanceLevelDescriptorObj.DescriptorId = SASOAPL.PerformanceLevelDescriptorId
+    WHERE(
+        StudentSchoolAssociation.ExitWithdrawDate IS NULL
+        OR StudentSchoolAssociation.ExitWithdrawDate >= GETDATE());
```
Comment (DavidJGapCR) [inline src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql:113]: We need to replace GETDATE() with NOW()

File: src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 6]
```diff
@@ -0,0 +1,113 @@
+﻿-- SPDX-License-Identifier: Apache-2.0
```
Comment (DavidJGapCR) [inline src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql:1]: If you need an ods in postgres there is one here: https://techdocs.ed-fi.org/display/EDFITOOLS/Testing+With+Sample+Data
Make sure AMT can be installed on Postgres.

**Comment by DavidJGapCR** [comment]
Since asmt_StudentAssessmentObjectiveFact doesn't exist anymore. We need to remove the attribute in line 40 of EdFi.AnalyticsMiddleTier.Tests.Operation.When_installing_Asmt_views

**Comment by DavidJGapCR** [comment]
I think StudentAssessmentObjectiveFact class can be removed, since we removed that View.

File: src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql (hunk 5 above)
Comment (NSD-Ruben) [reply src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql]: resolved, please review it again.

File: src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql (hunk 4 above)
Comment (NSD-Ruben) [reply src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql]: resolved, please review it again.

File: src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql (hunk 3 above)
Comment (NSD-Ruben) [reply src/EdFi.AnalyticsMiddleTier.DataStandard32/Asmt/PostgreSQL/0014-View-StudentAssessmentFact-Create.sql]: resolved, please review it again.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql (hunk 2 above)
Comment (NSD-Ruben) [reply src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql]: resolved, please review it again.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql (hunk 1 above)
Comment (NSD-Ruben) [reply src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql]: resolved, please review it again.

**Comment by DavidJGapCR** [comment]
Noticed that if I comment out some specific conditions in the joins, no test fails. These conditions are the ones in lines: 50,80 and 82.

File: src/EdFi.AnalyticsMiddleTier.Tests/TestCases/StudentAssessmentFact/0E886CBB_AB8F_ABD3_747F_C455C5141145_2009_04_01_100041249_should_have_PerformanceResult_empty.xml [hunk 7]
```diff
@@ -9,6 +9,5 @@
         WHERE StudentAssessmentKey = '0E886CBB-AB8F-ABD3-747F-C455C5141145-uri://edfi.org/Assessment-0E886CBB-AB8F-ABD3-747F-C455C5141145_2009-04-01-100041249';
     </Query>
     <Result>
-        <PerformanceResult></PerformanceResult>
```
Comment (DavidJGapCR) [inline src/EdFi.AnalyticsMiddleTier.Tests/TestCases/StudentAssessmentFact/0E886CBB_AB8F_ABD3_747F_C455C5141145_2009_04_01_100041249_should_have_PerformanceResult_empty.xml:12]: I think that what you are trying to say here is that it shouldn't return any results. I wouldn't say it is incorrect, but in other  cases we do something like 
`SELECT COUNT(1) AS CountValue`
And we have a CountResult class to map the result. You can keep it like this if you want but you would need to change the name of the test, so instead of _should_have_performanceresult_empty, it says that it returns no results.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+		CONCAT(Assessment.AssessmentIdentifier, '-', StudentAssessment.Namespace) AS AssessmentKey,
+        Assessment.AssessmentIdentifier,
+        StudentAssessment.Namespace,
+        StudentAssessment.StudentAssessmentIdentifier,
+        StudentAssessment.StudentUSI,
+        CONCAT(Student.StudentUniqueId, '-', StudentSchoolAssociation.SchoolId) AS StudentSchoolKey,
+        School.SchoolId AS SchoolKey,
+        CONVERT(varchar, StudentAssessment.AdministrationDate, 112) as AdministrationDate,
+        COALESCE(StudentAssessmentScoreResult.Result,SASOASR.Result) AS StudentScore,
+        COALESCE(ResultDatatypeTypeDescriptorDist.Description,ResultDescriptor.Description) AS ResultDataType,
```
Comment (DavidJGapCR) [inline src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:35]: I think I should have been more specific in my previous comment, sorry about that. 
But here, since ResultDatatypeTypeDescriptorDist.Description and ResultDescriptor.Description can be null, we should keep the empty result. Something like:
`COALESCE(ResultDatatypeTypeDescriptorDist.Description,ResultDescriptor.Description, '') AS ResultDataType,`
Same for the other COALESCE fields.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+BEGIN
+	DROP VIEW analytics.asmt_StudentAssessmentFact
+END
+GO
+
+CREATE VIEW analytics.asmt_StudentAssessmentFact AS
+
+	SELECT
+        CONCAT(StudentAssessment.AssessmentIdentifier, '-', StudentAssessment.Namespace, '-', StudentAssessment.StudentAssessmentIdentifier, '-', StudentAssessment.StudentUSI) AS StudentAssessmentKey,
+		CASE WHEN SASOA.StudentUSI IS NULL
```
Comment (DavidJGapCR) [inline src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:16]: Not related with this ticket. but this case statement makes me think that there could be a bug in **analytics.asmt_AssessmentFact**. Because there are concatenations of values that could be null. And there is no validation to check that. We could need a bug ticket to fix it.

File: src/EdFi.AnalyticsMiddleTier.Tests/TestCases/StudentAssessmentFact/0E886CBB_AB8F_ABD3_747F_C455C5141145_2009_04_01_100041249_should_have_PerformanceResult_empty.xml (hunk 7 above)
Comment (NSD-Ruben) [reply src/EdFi.AnalyticsMiddleTier.Tests/TestCases/StudentAssessmentFact/0E886CBB_AB8F_ABD3_747F_C455C5141145_2009_04_01_100041249_should_have_PerformanceResult_empty.xml:12]: resolved, please review it again.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql (hunk 8 above)
Comment (NSD-Ruben) [reply src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql]: resolved, please review it again.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+	SELECT
+        CONCAT(StudentAssessment.AssessmentIdentifier, '-', StudentAssessment.Namespace, '-', StudentAssessment.StudentAssessmentIdentifier, '-', StudentAssessment.StudentUSI) AS StudentAssessmentKey,
+		CASE WHEN SASOA.StudentUSI IS NULL 
+                THEN NULL 
+             ELSE 
+                CONCAT(SASOA.StudentUSI, '-', SASOA.IdentificationCode, '-', SASOA.AssessmentIdentifier, '-', SASOA.StudentAssessmentIdentifier, '-', SASOA.Namespace) 
+        END AS StudentObjectiveAssessmentKey,
+        CASE WHEN SASOA.AssessmentIdentifier IS NULL 
+                THEN NULL
```
Comment (jleiva-gap) [inline src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:22]: Replace Null with ''

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            ON Student.StudentUSI = StudentSchoolAssociation.StudentUSI
+    INNER JOIN
+        edfi.Descriptor AS EntryGradeLevelDescriptor
+            ON StudentSchoolAssociation.EntryGradeLevelDescriptorId = EntryGradeLevelDescriptor.DescriptorId
+    INNER JOIN
+        edfi.School
+            ON StudentSchoolAssociation.SchoolId = School.SchoolId
+    LEFT JOIN
+        edfi.StudentAssessmentScoreResult
+            ON StudentAssessment.AssessmentIdentifier = StudentAssessmentScoreResult.AssessmentIdentifier
```
Comment (jleiva-gap) [inline src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:58]: Test cases are missing to verify that this condition exists.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+            ON AssessmentReportingMethodDescriptor.AssessmentReportingMethodDescriptorId = AssessmentReportingMethodDescriptorDist.DescriptorId
+    LEFT JOIN
+        edfi.PerformanceLevelDescriptor
+            ON AssessmentPerformanceLevel.PerformanceLevelDescriptorId = PerformanceLevelDescriptor.PerformanceLevelDescriptorId
+    LEFT JOIN
+        edfi.Descriptor AS PerformanceLevelDescriptorDist
+            ON PerformanceLevelDescriptor.PerformanceLevelDescriptorId = PerformanceLevelDescriptorDist.DescriptorId
+	LEFT JOIN
+		edfi.StudentAssessmentStudentObjectiveAssessment SASOA
+			ON Assessment.AssessmentIdentifier = SASOA.AssessmentIdentifier
```
Comment (jleiva-gap) [inline src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:88]: Test cases are missing to verify that this condition exists.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+            AND 
+            Assessment.Namespace = SASOA.Namespace
```
Comment (jleiva-gap) [inline src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:90]: Test cases are missing to verify that this condition exists.

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+	DROP VIEW analytics.asmt_StudentAssessmentFact
+END
+GO
+
+CREATE VIEW analytics.asmt_StudentAssessmentFact AS
+
+	SELECT
+        CONCAT(StudentAssessment.AssessmentIdentifier, '-', StudentAssessment.Namespace, '-', StudentAssessment.StudentAssessmentIdentifier, '-', StudentAssessment.StudentUSI) AS StudentAssessmentKey,
+		CASE WHEN SASOA.StudentUSI IS NULL 
+                THEN NULL
```
Comment (jleiva-gap) [inline src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:17]: Replace Null with ''

**Review by jleiva-gap — CHANGES_REQUESTED** [review]

File: src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql (hunk 14 above)
Comment (jleiva-gap) [reply src/EdFi.AnalyticsMiddleTier.DataStandard31/Asmt/MSSQL/0014-View-StudentAssessmentFact-Create.sql:17]: Add a test case to validate the empty StudentObjectiveAssessmentKey and ObjectiveAssessmentKey.

**Review by jleiva-gap — CHANGES_REQUESTED** [review]

**Review by jleiva-gap — APPROVED** [review]
